### PR TITLE
Convolutional layer base implementation

### PR DESCRIPTION
**Pull request by seyounglee95**
This PR introduce new layer class `ConvolutionalLayer`. This layer extends `LayerBase`, and include implementation for `isLearnable()`, `getOutputShape()` and `generateParameterGradient()`. This PR include some modifications in `LayerParameter`, which enable `ConvolutionalLayer` to make use of its new parameters, `paddingHeight` and `paddingWidth`. Also, `ConvolutionalLayerConfigurationBuilder` class and `ConvolutionalLayerParameterInitializer` class are added for configuration of convolutional layer. `neural_network.proto` is modified to take inputs for convolutional layer. Concrete implementation for `feedForward()` and `backPropagate()` will be made in later PRs.

### CODE REVIEW

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/conf/ConvolutionalLayerConfigurationBuilder.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+import edu.snu.dolphin.dnn.layers.LayerBase;
+import edu.snu.dolphin.dnn.proto.NeuralNetworkProtos;
+import org.apache.reef.tang.Configuration;
+import org.apache.reef.tang.Tang;
+import org.apache.reef.util.Builder;
+
+/**
+ * Configuration builder for Convolutional layer.
+ *
+ * The configuration that this builder generates is used to create convolutional layer instance.
```
Comment (beomyeol) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/conf/ConvolutionalLayerConfigurationBuilder.java]: to create **a** convolutional layer instance.

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/conf/ConvolutionalLayerConfigurationBuilder.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+ * The generated configuration needs to bind the implementation for matrix factory and
+ * the parameter for a layer input shape, to inject a layer instance.
+ */
+public final class ConvolutionalLayerConfigurationBuilder implements Builder<Configuration> {
+
+  public static ConvolutionalLayerConfigurationBuilder newConfigurationBuilder() {
+    return new ConvolutionalLayerConfigurationBuilder();
+  }
+
+  private int paddingHeight;
```
Comment (beomyeol) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/conf/ConvolutionalLayerConfigurationBuilder.java]: Users do not have to specify `paddingHeight` when they use protocol buffer definition files. Shouldn't we set a default value for paddingHeight for consistency?
This is same for `paddingWidth`, `strideHeight`, and `strideWidth`.

If you agree with this, could you modify `PoolingLayerConfiguraitonBuilder`'s parameters that also have default values in the same way?

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layerparam/initializer/ConvolutionalLayerParameterInitializer.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+import javax.inject.Inject;
+
+import static edu.snu.dolphin.dnn.util.NeuralNetworkUtils.shapeFromString;
+
+/**
+ * Convolutional Layer parameter initializer.
+ *
+ * initializes the weight matrix with pseudo random normal distributed value with mean 0 and given standard deviation.
+ * initializes the bias vector with the given value.
+ * includes function that computes the output shape.
```
Comment (beomyeol) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layerparam/initializer/ConvolutionalLayerParameterInitializer.java]: Could you add subjects for these sentence?
If you're okay, please add subjects in `FullyConnectedLayerParameterInitializer`'s description, too.

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layerparam/initializer/ConvolutionalLayerParameterInitializer.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+   * @return the index of the layer.
+   */
+  public int getIndex() {
+    return index;
+  }
+
+  /**
+   * This function checks if stride is set proper for the input shape.
+   */
+  private void checkShape() {
```
Comment (beomyeol) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layerparam/initializer/ConvolutionalLayerParameterInitializer.java]: As mentioned in #169, Shouldn't we support inputs that do not satisfy this condition? Caffe supports such inputs.

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layers/ConvolutionalLayer.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+import javax.inject.Inject;
+
+  /**
+   * Convolutional layer.
+   *
+   * This layer is learnable having the updatable parameter (weight and bias).
+   * This layer works for only 2D inputs.
+   * In a forward pass,
+   * computes the product between weight and the input within kernel range and produce activation matrix.
```
Comment (beomyeol) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layers/ConvolutionalLayer.java]: No subject for this sentence.

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layers/ConvolutionalLayer.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+   * In a backward pass,
+   * error of each input pixel comes from
```
Comment (beomyeol) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layers/ConvolutionalLayer.java]: **the** error

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layers/ConvolutionalLayer.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+   * the product between weight and errors of output pixels affected by the input pixel in feedforward step.
+   */
```
Comment (beomyeol) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layers/ConvolutionalLayer.java]: indent

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layers/ConvolutionalLayer.java [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+  private final int paddingWidth;
+  private final int strideHeight;
+  private final int strideWidth;
+  private final int kernelHeight;
+  private final int kernelWidth;
+
+  @Inject
+  private ConvolutionalLayer(@Parameter(LayerIndex.class) final int index,
+                             @Parameter(LayerInputShape.class) final String inputShape,
+                             @Parameter(StrideHeight.class) final int paddingHeight,
```
Comment (beomyeol) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layers/ConvolutionalLayer.java]: `PaddingHeight.class`?

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layers/ConvolutionalLayer.java [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+                             @Parameter(StrideWidth.class) final int paddingWidth,
```
Comment (beomyeol) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layers/ConvolutionalLayer.java]: `PaddingWidth.class`?

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layers/ConvolutionalLayer.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+  @Override
+  public int[] getOutputShape() {
+    return outputShape;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public boolean isLearnable() {
+    return false;
```
Comment (beomyeol) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layers/ConvolutionalLayer.java]: This should be `true`.

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layers/ConvolutionalLayer.java [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+   */
+  @Override
+  public Matrix backPropagate(final Matrix input, final Matrix activation, final Matrix nextError) {
+    throw new RuntimeException("Not Implemented");
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public LayerParameter generateParameterGradient(final Matrix input, final Matrix error) {
+    return LayerParameter.newBuilder()
```
Comment (beomyeol) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layers/ConvolutionalLayer.java]: The method to generate `ConvolutionalLayer`'s parameter gradient is different from `FullyConnectedLayer`'s one, isn't it? I referred to [this webpage](http://ufldl.stanford.edu/tutorial/supervised/ConvolutionalNeuralNetwork).

**Comment by beomyeol** [comment]
@seyounglee95 I've done my pass. Please check comments that I left. :)

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layers/ConvolutionalLayer.java (hunk 11 above)
Comment (seyounglee95) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layers/ConvolutionalLayer.java]: I'll implement it in another PR.

**Comment by seyounglee95** [comment]
@beomyeol Thank you for your comments. I pushed a commit. Please take a look.

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layerparam/initializer/ConvolutionalLayerParameterInitializer.java [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+   * This function computes output shape.
+   * input shape: row * col
+   * output shape: row' * col'
+   * row' = (row − kernelHeight + 2 * paddingHeight) / stride + 1
+   * col' = (col − kernelWidth + 2 * paddingWidth) / stride + 1
+   * @return shape of output
+   */
+  private int[] computeOutputShape() {
+    final int[] computedShape = new int[2];
+    computedShape[0] = (int) Math.ceil((double) (inputShape[0] - kernelHeight + 2 * paddingHeight) / strideHeight) + 1;
```
Comment (beomyeol) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layerparam/initializer/ConvolutionalLayerParameterInitializer.java]: How about changing `double` to `float` like `PoolingLayer`?

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layerparam/initializer/ConvolutionalLayerParameterInitializer.java [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+   */
+  public int getIndex() {
+    return index;
+  }
+
+  /**
+   * This function computes output shape.
+   * input shape: row * col
+   * output shape: row' * col'
+   * row' = (row − kernelHeight + 2 * paddingHeight) / stride + 1
```
Comment (beomyeol) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layerparam/initializer/ConvolutionalLayerParameterInitializer.java]: The description about `row'` and `col`' is obsolete. Can you fix the description for `PoolingLayerParameterInitializer`, too?

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layerparam/initializer/ConvolutionalLayerParameterInitializer.java (hunk 12 above)
Comment (beomyeol) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layerparam/initializer/ConvolutionalLayerParameterInitializer.java]: In addition, please add the code that an exception will be thrown when the input shape is not 2D.

**Comment by beomyeol** [comment]
@seyounglee95 I've done my pass and left some comments. I'll merge this after they are resolved.

File: dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layerparam/initializer/ConvolutionalLayerParameterInitializer.java (hunk 13 above)
Comment (seyounglee95) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layerparam/initializer/ConvolutionalLayerParameterInitializer.java]: I'll modify it in another PR. I'm going to make PR about implementation of padding parameters on pooling layers soon.

Comment (beomyeol) [inline dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layerparam/initializer/ConvolutionalLayerParameterInitializer.java]: That's okay. Please don't forget to include the modification for this. :)

**Comment by seyounglee95** [comment]
@beomyeol Thank you for your comments. Please take a look.

**Comment by beomyeol** [comment]
@seyounglee95 This looks good. I'm merging.

**Pull request merged**